File: Archive_LJ_isothermal_isobaric/gnptmclj.f90
!**********************************************************************!
!
! File: gnptmclj.f90
!
! Create random ("gas") initial configuration for NpT-Monte Carlo of
! Lennard-Jonesium
!
! 29-Apr-1999 (MN)
! 17-Apr-2012
!
!**********************************************************************!

program gnptmclj

!**********************************************************************!

  implicit none

  integer,parameter::iout=2

  character(len=80)::fname
  integer::n,nt,ntjob,ntprint,ntskip
  integer::i,m,nran
  integer,dimension(:),allocatable::iran
  real::disp,dlnv,dr,p,pvm,t,v
  real::c,ran,rho
  real,dimension(:),allocatable::x,y,z

! User input

  magic: do
    write(unit=*,fmt="(a)",advance="no") "              n="
    read(unit=*,fmt=*) n
    m=0
    do
      m=m+2
      if(m**3==2*n) then     ! Check if magic number
        exit magic
      else if(m**3>2*n) then
        exit
      end if
    end do
  end do magic

  write(unit=*,fmt="(a)",advance="no") "              p="
  read(unit=*,fmt=*) p
  write(unit=*,fmt="(a)",advance="no") "              t="
  read(unit=*,fmt=*) t
  write(unit=*,fmt="(a)",advance="no") "            rho="
  read(unit=*,fmt=*) rho
  write(unit=*,fmt="(a)",advance="no") "           disp="
  read(unit=*,fmt=*) disp
  write(unit=*,fmt="(a)",advance="no") "           dv/v="
  read(unit=*,fmt=*) dlnv
  write(unit=*,fmt="(a)",advance="no") "    prob(vmove)="
  read(unit=*,fmt=*) pvm
  write(unit=*,fmt="(a)",advance="no") "             dr="
  read(unit=*,fmt=*) dr
  write(unit=*,fmt="(a)",advance="no") "         ntskip="
  read(unit=*,fmt=*) ntskip
  write(unit=*,fmt="(a)",advance="no") " ntprint/ntskip="
  read(unit=*,fmt=*) ntprint
  write(unit=*,fmt="(a)",advance="no") "   ntjob/ntskip="
  read(unit=*,fmt=*) ntjob
  write(unit=*,fmt="(a)",advance="no") &
       "          fname=[nptmclj_in.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="nptmclj_in.dat"
  end if

! RNG characteristics

  call random_seed(size=nran)

! Allocate arrays

  allocate(iran(nran),x(n),y(n),z(n))

! Random positions

  v=n/rho
  c=v**(1.0/3.0)
  do i=1,n
    call random_number(ran)
    x(i)=(ran-0.5)*c
    call random_number(ran)
    y(i)=(ran-0.5)*c
    call random_number(ran)
    z(i)=(ran-0.5)*c
  end do

! RNG seed

  call random_seed(get=iran)

! Write startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dlnv,dr,p,pvm,t,v
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program gnptmclj

!**********************************************************************!

File: Archive_LJ_isothermal_isobaric/gnptmclj.c
/**********************************************************************
 *
 * File: gnptmclj.c
 *
 * Create random ("gas") initial configuration for NpT-Monte Carlo of
 * Lennard-Jonesium
 *
 * 03-Apr-2010 (MN)
 * 17-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.c"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpo;
  char fname[BSIZE];
  unsigned short dummy[3];
  unsigned short *seed;
  int n,nt,ntjob,ntprint,ntskip;
  int i,m;
  double disp,dlnv,dr,p,pvm,t,v;
  double c,rho;
  double *x,*y,*z;

  /* User input */

  for(;;) {
    printf("              n=");
    getval_i(&n);
    m=0;
    while(m*m*m<2*n) {
      m=m+2;
      if(m*m*m==2*n) { /* Check if magic number */
        goto magic;
      }
    }
  }
  magic:

  printf("              p=");
  getval_d(&p);
  printf("              t=");
  getval_d(&t);
  printf("            rho=");
  getval_d(&rho);
  printf("           disp=");
  getval_d(&disp);
  printf("           dv/v=");
  getval_d(&dlnv);
  printf("    prob(vmove)=");
  getval_d(&pvm);
  printf("             dr=");
  getval_d(&dr);
  printf("         ntskip=");
  getval_i(&ntskip);
  printf(" ntprint/ntskip=");
  getval_i(&ntprint);
  printf("   ntjob/ntskip=");
  getval_i(&ntjob);

  strcpy(fname,"nptmclj_in.dat");
  printf("          fname=[nptmclj_i.dat] ");
  getval_s(fname);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Random positions */

  v=n/rho;
  c=cbrt(v);
  for(i=0;i<=n-1;i++) {
    x[i]=c*(drand48()-0.5);
    y[i]=c*(drand48()-0.5);
    z[i]=c*(drand48()-0.5);
  }

  /* RNG seed */

  seed=seed48(dummy);

  /* Write startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(&seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dlnv,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&p,sizeof(double),1,fpo);
  fwrite(&pvm,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);
  fwrite(&v,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/
